kopf/reactor/daemons.py

```diff
@@ -0,0 +1,394 @@
+                stopper=stopper,  # for checking (passed to kwargs)
+            )
+            daemon = containers.Daemon(
+                stopper=stopper,  # for stopping (outside of causes)
+                handler=handler,
+                logger=logging_engine.LocalObjectLogger(body=cause.body),
+                task=asyncio.create_task(_runner(
+                    handler=handler,
+                    cause=daemon_cause,
+                    memory=memory,
+                )),
+            )
+            daemons[daemon_id] = daemon
+    return []
+
+
+async def stop_resource_daemons(
+        *,
+        daemons: MutableMapping[containers.DaemonId, containers.Daemon],
+) -> Collection[float]:
+    """
+    Terminate all daemons of an individual resource (gracefully and by force).
+
+    All daemons are terminated in parallel to speed up the termination
+    (especially taking into account that some daemons can take time to finish).
+
+    The daemons are asked to terminate as soon as the object is marked
+    for deletion. It can take some time until the deletion handlers also
+    finish their work. The object is not physically deleted until all
+    the daemons are terminated (by putting a finalizer on it).
+    """
+    delays: List[float] = []
+    now = time.monotonic()
+    for daemon_id, daemon in daemons.items():
+        logger = daemon.logger
+        stopper = daemon.stopper
+        age = (now - (stopper.when or now))
+
+        if isinstance(daemon.handler, handlers_.ResourceDaemonHandler):
+            backoff = daemon.handler.cancellation_backoff
+            timeout = daemon.handler.cancellation_timeout
+            polling = daemon.handler.cancellation_polling or DAEMON_POLLING_INTERVAL
+        elif isinstance(daemon.handler, handlers_.ResourceTimerHandler):
+            backoff = None
+            timeout = None
+            polling = DAEMON_POLLING_INTERVAL
+        else:
+            raise RuntimeError(f"Unsupported daemon handler: {daemon.handler!r}")
+
+        # Whatever happens with other flags & logs & timings, this flag must be surely set.
+        stopper.set(reason=primitives.DaemonStoppingReason.RESOURCE_DELETED)
+
+        # It might be so, that the daemon exits instantly (if written properly). Avoid patching and
+        # unnecessary handling cycles in this case: just give the asyncio event loop an extra cycle.
+        await asyncio.sleep(0)
+
+        # Try different approaches to exiting the daemon based on timings.
+        if daemon.task.done():
+            pass
+
+        elif backoff is not None and age < backoff:
+            if not stopper.is_set(reason=primitives.DaemonStoppingReason.DAEMON_SIGNALLED):
+                stopper.set(reason=primitives.DaemonStoppingReason.DAEMON_SIGNALLED)
+                logger.debug(f"Daemon {daemon_id!r} is signalled to exit gracefully.")
+            delays.append(backoff - age)
+
+        elif timeout is not None and age < timeout + (backoff or 0):
```

Comment:
should it be rather `timeout - (backoff or 0)` here ?

Reply:
Indeed. Fixed.

Reply:
It is correct as it is. Here, the `age` is measured since the termination itself has started. However, the `cancellation_timeout` should be measured since `task.cancel()`. Which means, the real age limit is `backoff` + `cancellation_timeout` — unlike the waiting timeout below for sleep-or-wait.